contracts/contracts/strategies/NativeStaking/CompoundingValidatorManager.sol

```diff
@@ -0,0 +1,904 @@
+        // the deposit in `stakeETH` was created on the execution layer.
+        require(
+            deposit.blockNumber <= depositBlockNumber,
+            "Deposit block before deposit"
+        );
+
+        // Get the slot that is on or after the deposit was made on the execution layer.
+        uint64 depositSlot = IBeaconOracle(BEACON_ORACLE).blockToSlot(
+            depositBlockNumber
+        );
+
+        // Check the deposit slot is before the first pending deposit's slot on the beacon chain.
+        // If this is not true then we can't guarantee the deposit has been processed by the beacon chain.
+        // The deposit's slot can not be the same slot as the first pending deposit as there could be
+        // many deposits in the same block, hence have the same pending deposit slot.
+        // If the first pending deposit's slot is 0 then there are no pending deposits so
+        // our deposit must have been processed on the beacon chain.
+        require(
+            depositSlot < firstPendingDepositSlot ||
+                firstPendingDepositSlot == 0,
+            "Deposit not processed"
+        );
+
+        // The slot of the execution layer deposit must be before
+        // the slot that the deposit was processed on the beacon chain.
+        require(depositSlot < processedSlot, "Slot not after deposit");
+
+        // Get the beacon block root for the slot that is on or after the deposit was processed on the beacon chain.
+        bytes32 blockRoot = IBeaconOracle(BEACON_ORACLE).slotToRoot(
+            processedSlot
+        );
+
+        // Verify the slot of the first pending deposit matches the beacon chain
+        IBeaconProofs(BEACON_PROOFS).verifyFirstPendingDepositSlot(
+            blockRoot,
+            firstPendingDepositSlot,
+            firstPendingDepositSlotProof
+        );
+
+        // After verifying the proof, update the contract storage
+        deposits[depositDataRoot].status = DepositStatus.VERIFIED;
+        // Move the last deposit to the index of the verified deposit
+        bytes32 lastDepositDataRoot = depositsRoots[depositsRoots.length - 1];
+        depositsRoots[deposit.depositIndex] = lastDepositDataRoot;
+        deposits[lastDepositDataRoot].depositIndex = deposit.depositIndex;
+        // Delete the last deposit from the list
+        depositsRoots.pop();
+
+        emit DepositVerified(
+            depositDataRoot,
+            uint256(deposit.amountGwei) * 1 gwei
+        );
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    // TODO what if the last validator was exited rather than consolidated?
+    // slither-disable-start reentrancy-no-eth
+    function verifyConsolidation(
+        uint64 parentBlockTimestamp,
+        uint64 lastValidatorIndex,
+        bytes calldata validatorPubKeyProof,
+        bytes32 balancesLeaf,
+        bytes calldata validatorBalanceProof
+    ) external onlyRegistrator {
```

Comment:
It would seem like clean symmetrical code to have this also check the contract is paused before allowing a call to it.  (The ValidatorRegistrar does do this this check)